File: CPE/Exam/231212/H7_UVa10721.py
from functools import *

"""
    DP
    n 個units, k 個bars, 每個bar最多m個units
    第一個unit必須是black
"""

@cache
def BC(n, k, m):
    if n == 0: # 沒有unit了，若還有bar則不滿足
        return 1 if k == 0 else 0
    if k == 1: # 只有1個bar，若滿足限制條件則有1種可能
        return 1 if n <= m else 0
    res = 0
    for x in range(1, min(n, m)+1):
        res += BC(n-x, k-1, m) # 剩下n-x個units, k-1個bars, 每個bar最多還是m個units
    return res

while True:
    try:
        n, k, m = map(int, input().split())  
    except EOFError:
        break
    print(BC(n, k, m))

File: CPE/Exam/231212/H7_UVa10721.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const int N = 55;
#define endl '\n'

LL dp[N][N][N];

LL BC(int n, int k, int m) {
    if (dp[n][k][m] != -1) return dp[n][k][m]; // memoization
    if (n == 0) return k == 0; // 沒有unit了，若還有bar則不滿足
    if (k == 1) return n <= m; // 只有1個bar，若滿足限制條件則有1種可能
    LL res = 0;
    for (int x = 1; x <= min(n, m); ++x) {
        res += BC(n-x, k-1, m); // 剩下n-x個units, k-1個bars, 每個bar最多還是m個units
    }
    dp[n][k][m] = res;
    return res;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, k, m;
    memset(dp, -1, sizeof(dp));
    while (cin >> n >> k >> m) {
        cout << BC(n, k, m) << endl;
    }
    return 0;
}